controllers/storagecluster/exporter.go

```diff
@@ -209,9 +265,515 @@ func CreateOrUpdateServiceMonitor(r *StorageClusterReconciler, instance *ocsv1.S
+	r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+	configMapName := "ocs-metrics-exporter-ceph-conf"
+	currentConfigMap := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      configMapName,
+			Namespace: instance.Namespace,
+		},
+	}
+	_, err := controllerutil.CreateOrUpdate(ctx, r.Client, currentConfigMap, func() error {
+		currentConfigMap.Data = map[string]string{
+			"ceph.conf": `
+[global]
+auth_cluster_required = cephx
+auth_service_required = cephx
+auth_client_required = cephx
+`,
+			// keyring is a required key and its value should be empty
+			"keyring": "",
+		}
+		return nil
+	})
+	return err
+}
+
+const metricsExporterClusterRoleJSON = `
+{
+	"apiVersion":"rbac.authorization.k8s.io/v1",
+	"kind":"ClusterRole",
+	"metadata":{"name":"ocs-metrics-exporter"},
+	"rules":[
+		{
+			"apiGroups":["monitoring.coreos.com"],
+			"resources":["*"],
+			"verbs":["*"]
+		},
+		{
+			"apiGroups":[""],
+			"resources":["persistentvolumes","persistentvolumeclaims","nodes","pods","configmaps","secrets"],
+			"verbs":["get","list","watch"]
+		},
+		{
+			"apiGroups":["ceph.rook.io"],
+			"resources":["cephobjectstores","cephclusters","cephblockpools","cephrbdmirrors"],
+			"verbs":["get","list","watch"]
+		},
+		{
+			"apiGroups":["quota.openshift.io"],
+			"resources":["clusterresourcequotas"],
+			"verbs":["get","list","watch"]
+		},
+		{
+			"apiGroups":["objectbucket.io"],
+			"resources":["objectbuckets","objectbucketclaims"],
+			"verbs":["get","list"]
+		},
+		{
+			"apiGroups":[""],
+			"resources":["secrets"],
+			"verbs":["get","list"]
+		},
+		{
+			"apiGroups":["storage.k8s.io"],
+			"resources":["storageclasses"],
+			"verbs":["get","list","watch"]
+		},
+		{
+			"apiGroups":["ocs.openshift.io"],
+			"resources":["storageconsumers","storageclusters"],
+			"verbs":["get","list","watch"]
+		}
+	]
+}`
+
+func updateMetricsExporterClusterRoles(ctx context.Context,
+	r *StorageClusterReconciler, instance *ocsv1.StorageCluster) error {
+	currentClusterRole := new(rbacv1.ClusterRole)
+	var expectedClusterRole = new(rbacv1.ClusterRole)
+	err := json.Unmarshal([]byte(metricsExporterClusterRoleJSON), expectedClusterRole)
+	if err != nil {
+		r.Log.Error(err, "an unexpected error occurred while unmarshalling clusterrole yaml")
+		return err
+	}
+	expectedClusterRole.Name, currentClusterRole.Name = metricsExporterName, metricsExporterName
+	_, err = controllerutil.CreateOrUpdate(ctx, r.Client, currentClusterRole, func() error {
+		expectedRulesLen := len(expectedClusterRole.Rules)
+		if len(currentClusterRole.Rules) != expectedRulesLen {
+			currentClusterRole.Rules = make([]rbacv1.PolicyRule, expectedRulesLen)
+		}
+		copy(currentClusterRole.Rules, expectedClusterRole.Rules)
+		return nil
+	})
+	return err
+}
+
+const expectedMetricExporterRoleJSON = `
+{
+	"apiVersion":"rbac.authorization.k8s.io/v1",
+	"kind":"Role",
+	"metadata":{"name":"ocs-metrics-svc"},
```

Comment:
`Role` is namespaced resource. Add namespace.

Reply:
All the json string variables are static templates. Needed dynamic parts are added in the code.